DSpace/DSpace
# Simplify process of fixing the tests after adding sidebar facets

## Description
When you add a new sidebar facet to the `defaultConfiguration` in `discovery.xml` the `DiscoveryRestControllerIT` tests fail. To fix this you need to create a new `FacetEntryMatcher` & `SearchFilterMatcher` and afterwards you need to add the `FacetEntryMatcher` on 30 different places in `DiscoveryRestControllerIT`. This PR simplifies this process by creating an array in which you can add `FacetEntryMatcher`s that will automatically be added on those 30 places. I also already created the same config for the `SearchFilterMatcher` & `SortOptionMatcher` to have all those fields grouped together at the top of the file.

## Instructions for Reviewers
List of changes in this PR:
* Created 3 lists in which you can add the `FacetEntryMatcher`, `SearchFilterMatcher` & `SortOptionMatcher` that you added to the `defaultConfiguration` bean in `discovery.xml`.
* Refactored the tests to always add these matchers to their expected result

**Guidance for how to test/review this PR:**
- Add a new sidebar facet to the `defaultConfiguration` in `discovery.xml` for example:
   ```xml
    <bean id="searchFilterSerialNumber" class="org.dspace.discovery.configuration.HierarchicalSidebarFacetConfiguration">
        <property name="indexFieldName" value="serialNumber"/>
        <property name="metadataFields">
            <list>
                <value>dc.identifier</value>
            </list>
        </property>
        <property name="facetLimit" value="5"/>
        <property name="sortOrderSidebar" value="COUNT"/>
        <property name="sortOrderFilterPage" value="COUNT"/>
        <property name="splitter" value="::"/>
    </bean>
   ```
- Add this bean to the `sidebarFacets` & `searchFilters` array in `discovery.xml`
- Create their matchers and add them to the `customSidebarFacets` & `customSearchFilters`, and check that all the test suceed without having to add the matcher at those 30 places
  ```java
    public static Matcher<? super Object> serialNumberFacet(boolean hasNext) {
        return allOf(
            hasJsonPath("$.name", is("serialNumber")),
            hasJsonPath("$.facetLimit", any(Integer.class)),
            hasJsonPath("$._links.self.href", containsString("api/discover/facets/serialNumber")),
            hasJsonPath("$._links", matchNextLink(hasNext, "api/discover/facets/serialNumber"))
        );
    }
  ```
  ```java
    public static Matcher<? super Object> serialNumberFilter() {
        return allOf(
            hasJsonPath("$.filter", is("serialNumber")),
            hasJsonPath("$.hasFacets", is(true)),
            hasJsonPath("$.type", is("hierarchical")),
            hasJsonPath("$.openByDefault", is(false)),
            checkOperators()
        );
    }
  ```

## Checklist
